src/codecs/avif/decoder.rs

```diff
@@ -0,0 +1,156 @@
+//! Decoding of AVIF images.
+///
+/// The [AVIF] specification defines an image derivative of the AV1 bitstream, an open video codec.
+///
+/// [AVIF]: https://aomediacodec.github.io/av1-avif/
+
+use std::convert::TryFrom;
+use std::io::{self, Cursor, Read};
+use std::marker::PhantomData;
+use std::mem;
+
+use crate::{ColorType, ImageResult, ImageDecoder};
+
+use mp4parse::read_avif;
+use dav1d::{PlanarImageComponent, PixelLayout};
+
+/// AVIF Decoder.
+///
+/// Reads one image into the chosen input.
+pub struct AvifDecoder<R> {
+    inner: PhantomData<R>,
+    picture: dav1d::Picture,
+    alpha_picture: Option<dav1d::Picture>,
+}
+
+impl<R: Read> AvifDecoder<R> {
+    /// Create a new decoder that reads its input from `r`.
+    pub fn new(mut r: R) -> ImageResult<Self> {
+        let ctx = read_avif(&mut r).unwrap();
+        let mut primary_decoder = dav1d::Decoder::new();
+        primary_decoder.send_data(ctx.primary_item(), None, None, None).unwrap();
+        let picture = primary_decoder.get_picture().unwrap();
+        let alpha_picture = if let Some(alpha_item) = ctx.alpha_item() {
+            let mut alpha_decoder = dav1d::Decoder::new();
+            alpha_decoder.send_data(alpha_item, None, None, None).unwrap();
+            Some(alpha_decoder.get_picture().unwrap())
+        } else {
+            None
+        };
```

Comment:
This should not `unwrap` nor `assert`, it handles potentially dangerous user input.

Reply:
Once https://github.com/rust-av/dav1d-rs/pull/42 is merged, this won’t have any `unwrap` anymore (see the current draft version).